src/System.Globalization.Calendars/tests/CalendarHelpers.cs

```diff
@@ -0,0 +1,185 @@
+        {
+            return calendar.GetEra(calendar.MaxSupportedDateTime);
+        }
+
+        private static int YearOffset(Calendar calendar)
+        {
+            if (calendar is KoreanCalendar)
+            {
+                return 2333;
+            }
+            if (calendar is ThaiBuddhistCalendar)
+            {
+                return 543;
+            }
+            return 0;
+        }
+
+        public static void AddYears_Invalid(Calendar calendar)
+        {
+            Assert.Throws<ArgumentOutOfRangeException>(() => calendar.AddYears(calendar.MaxSupportedDateTime, 1));
+            Assert.Throws<ArgumentOutOfRangeException>(() => calendar.AddYears(calendar.MinSupportedDateTime, -1));
+        }
+
+        public static void AddMonths_Invalid(Calendar calendar)
+        {
+            Assert.Throws<ArgumentOutOfRangeException>(() => calendar.AddMonths(calendar.MaxSupportedDateTime, 1));
+            Assert.Throws<ArgumentOutOfRangeException>(() => calendar.AddMonths(calendar.MinSupportedDateTime, -1));
+
+            Assert.Throws<ArgumentOutOfRangeException>("months", () => calendar.AddMonths(DateTime.Now, -120001));
+            Assert.Throws<ArgumentOutOfRangeException>("months", () => calendar.AddMonths(DateTime.Now, 120001));
+        }
+
+        public static void GetDaysInYear_Invalid(Calendar calendar)
+        {
+            int yearOffset = YearOffset(calendar);
+            int era = MaxEra(calendar);
+
+            // Year is outside supported range
+            Assert.Throws<ArgumentOutOfRangeException>("year", () => calendar.GetDaysInYear(-1, era));
+            Assert.Throws<ArgumentOutOfRangeException>("year", () => calendar.GetDaysInYear(calendar.MinSupportedDateTime.Year + yearOffset - 1, era));
+            Assert.Throws<ArgumentOutOfRangeException>("year", () => calendar.GetDaysInYear(calendar.MaxSupportedDateTime.Year + yearOffset + 1, era));
+
+            // Era is outside supported range
+            Assert.Throws<ArgumentOutOfRangeException>("era", () => calendar.GetDaysInYear(yearOffset + 1, -era));
+            Assert.Throws<ArgumentOutOfRangeException>("era", () => calendar.GetDaysInYear(yearOffset + 1, era + 1));
+        }
+
+        public static void GetDaysInMonth_Invalid(Calendar calendar)
+        {
+            int yearOffset = YearOffset(calendar);
+            int era = MaxEra(calendar);
+
+            // Year is outside supported range
+            Assert.Throws<ArgumentOutOfRangeException>("year", () => calendar.GetDaysInMonth(-1, 10, era));
+            Assert.Throws<ArgumentOutOfRangeException>("year", () => calendar.GetDaysInMonth(0, 10, era));
+            Assert.Throws<ArgumentOutOfRangeException>("year", () => calendar.GetDaysInMonth(calendar.MinSupportedDateTime.Year + yearOffset - 1, 10, era));
+            Assert.Throws<ArgumentOutOfRangeException>("year", () => calendar.GetDaysInMonth(calendar.MaxSupportedDateTime.Year + yearOffset + 1, 10, era));
```

Comment:
Here ```era``` is set to the max era, but valid years change based on the era.  For the ThaiBuddistCalendar, there's only one era so it doesn't matter, but for example for the JapaneseCalendar there are four eras, and there it does matter, e.g. this code:
```C#
var c = new JapaneseCalendar();
foreach (int era in c.Eras)
{
    try { c.GetDaysInYear(int.MaxValue, era); } catch (Exception e) { Console.WriteLine(e.Message); }
}
```
prints:
```
Valid values are between 1 and 8011, inclusive.
Parameter name: year
Valid values are between 1 and 64, inclusive.
Parameter name: year
Valid values are between 1 and 15, inclusive.
Parameter name: year
Valid values are between 1 and 45, inclusive.
Parameter name: year
```